cmd/launchBackup.go

```diff
@@ -0,0 +1,234 @@
+
+	// check for back slash in the BackupPath
+	if check := strings.Contains(BackupPath[len(BackupPath)-1:], "/"); check {
+		BackupPath = BackupPath[:len(BackupPath)-1]
+	}
+
+	err := cleanup(BackupPath)
+	if err != nil {
+		log.Error(fmt.Printf("Old directories couldn't be deleted, err: %s\n", err))
+	}
+
+	log.Info("Old contents have been cleaned up")
+
+	path, err := createDir(BackupPath)
+	if err != nil {
+		log.Error(fmt.Printf("Old directories couldn't be deleted, err: %s\n", err))
+	}
+
+	//ostree pinning
+	ostreeArgs := "ostree admin pin 0"
+	err = executeArgs(ostreeArgs, "backup-time", "ostree")
+	if err != nil {
+		log.Error(err)
+	}
+
+	// for now we skip the container image backup
+	/*	// container image backup
+		bashArgs := fmt.Sprintf("for id in $(crictl images -o json | jq -r '.images[].id'); do mkdir -p %s/$id; /usr/bin/skopeo copy --all --insecure-policy containers-storage:$id dir:%s/$id; done", path[0], path[0])
+		err = executeArgs(bashArgs, path[0], "containers")
+		if err != nil {
+			log.Warn(err)
+		}
+	*/
+	// cluster backup
+	err = executeArgs(BackupScript, path[1], "etcd-cluster")
+	if err != nil {
+		return err
+	}
+
+	// etc back up
+	etcExcludeArgs := fmt.Sprintf("cat /etc/tmpfiles.d/* | sed 's/#.*//' | awk '{print $2}' | grep '^/etc/' | sed 's#^/etc/##' > %s/etc.exclude.list; echo '.updated' >> %s/etc.exclude.list; echo 'kubernetes/manifests' >> %s/etc.exclude.list", BackupPath, BackupPath, BackupPath)
+	err = executeArgs(etcExcludeArgs, BackupPath, "etc-exclude-list")
+	if err != nil {
+		log.Error(err)
+	}
+
+	// usrlocal backup
+	etcArgs := fmt.Sprintf("rsync -a %s %s", etc, path[2])
+	err = executeArgs(etcArgs, path[2], etc)
+	if err != nil {
+		log.Error(err)
+	}
+
+	usrlocalArgs := fmt.Sprintf("rsync -a %s %s", usrlocal, path[3])
+	err = executeArgs(usrlocalArgs, path[3], usrlocal)
+	if err != nil {
+		log.Error(err)
+	}
+
+	kubeletArgs := fmt.Sprintf("rsync -a %s %s", kubelet, path[4])
+	err = executeArgs(kubeletArgs, path[4], kubelet)
+	if err != nil {
+		log.Error(err)
+	}
+
+	log.Info(strings.Repeat("-", 60))
+	log.Info("backup has succesfully finished ...")
+
+	return nil
+
+}
+
+// cleanup deletes all old sub diectories and files in the recovery partition
+// returns: 			error
+func cleanup(path string) error {
+	//change root directory to /host
+	if err := syscall.Chroot(host); err != nil {
+		log.Println("Couldn't do chroot to %s", host)
+		return nil
+	}
+
+	log.Info(strings.Repeat("-", 60))
+	log.Info("Cleaning up old contents and ostree deployment started ...")
+	log.Info(strings.Repeat("-", 60))
+	// Cleanup previous backups
+	dir, _ := os.Open(path)
+	subDir, _ := dir.Readdir(0)
+
+	// Loop over the directory's files.
+	for index := range subDir {
+		fileNames := subDir[index]
+
+		// Get name of file and its full path.
+		name := fileNames.Name()
+		fullPath := path + "/" + name
+		log.Info("\nfullpath: ", fullPath)
+
+		if subDir[index].Name() != "extras.tgz" {
```

Comment:
I don't see a step to create the extras.tgz, but why is this check here to avoid its deletion?

Reply:
ideally i can't take the backup using OC in this workflow, it requires the path of kubeconfig and launching a OC command from the container may not be a good idea. We have to see how we can take backup with below commands. so the initially i thought may be we can keep that somehow before we execute the backup.

oc get mc -o=jsonpath='{range .items[*]}{range .spec.config.storage.files[*]}{.path}{"\n"}' \
    | sort -u | grep -v -e '^/etc/' -e '^/usr/local/' -e '^$' \
    | xargs --no-run-if-empty tar czf /var/recovery/extras.tgz

Reply:
I think that could be an issue here, where we could have a stale extras.tgz, or pretty much anything there. I'd rather just purge whatever is existing in /var/recovery. If we find we need a workaround for extras.tgz, we can always create it manually to a different location before starting the upgrade, and copy it into /var/recovery prior to running the recovery. For that matter, if we don't have a safe way to create the file, I'd rather just drop the support from the recovery tool

Reply:
@donpenney Removed as suggested in commit 7b894eaeb1a0f3a62ea17f843b347cdd677740a2